commons-core/src/main/scala/com/avsystem/commons/rest/RawRest.scala

```diff
@@ -0,0 +1,177 @@
+
+  def rpcChainRepr: String =
+    prefixes.iterator.map(_.rpcName).mkString("", "->", s"->${finalCall.rpcName}")
+}
+
+@methodTag[RestMethodTag]
+trait RawRest {
+
+  import RawRest._
+
+  @multi
+  @tried
+  @tagged[Prefix](whenUntagged = new Prefix)
+  @paramTag[RestParamTag](defaultTag = new Path)
+  def prefix(@methodName name: String, @composite headers: RestHeaders): Try[RawRest]
+
+  @multi
+  @tried
+  @tagged[GET]
+  @paramTag[RestParamTag](defaultTag = new Query)
+  def get(@methodName name: String, @composite headers: RestHeaders): Async[RestResponse]
+
+  @multi
+  @tried
+  @tagged[BodyMethodTag](whenUntagged = new POST)
+  @paramTag[RestParamTag](defaultTag = new JsonBodyParam)
+  def handle(@methodName name: String, @composite headers: RestHeaders,
+    @multi @tagged[JsonBodyParam] body: NamedParams[JsonValue]): Async[RestResponse]
+
+  @multi
+  @tried
+  @tagged[BodyMethodTag](whenUntagged = new POST)
+  @paramTag[RestParamTag]
+  def handleSingle(@methodName name: String, @composite headers: RestHeaders,
+    @encoded @tagged[Body] body: HttpBody): Async[RestResponse]
+
+  def asHandleRequest(metadata: RestMetadata[_]): HandleRequest = {
+    metadata.ensureUnambiguousPaths()
+    metadata.ensureUniqueParams(Nil)
+    RawRest.safeHandle { case RestRequest(method, headers, body) =>
+      metadata.resolvePath(method, headers.path).toList match {
+        case List(ResolvedPath(prefixes, RestMethodCall(finalRpcName, finalPathParams, _), singleBody)) =>
+          def resolveCall(rawRest: RawRest, prefixes: List[RestMethodCall]): Async[RestResponse] = prefixes match {
+            case RestMethodCall(rpcName, pathParams, _) :: tail =>
+              rawRest.prefix(rpcName, headers.copy(path = pathParams)) match {
+                case Success(nextRawRest) => resolveCall(nextRawRest, tail)
+                case Failure(e: HttpErrorException) => RawRest.successfulAsync(e.toResponse)
+                case Failure(cause) => RawRest.failingAsync(cause)
+              }
+            case Nil =>
+              val finalHeaders = headers.copy(path = finalPathParams)
+              if (method == HttpMethod.GET)
+                rawRest.get(finalRpcName, finalHeaders)
+              else if (singleBody)
+                rawRest.handleSingle(finalRpcName, finalHeaders, body)
+              else
+                rawRest.handle(finalRpcName, finalHeaders, HttpBody.parseJsonBody(body))
+          }
+          resolveCall(this, prefixes)
+        case Nil =>
+          val pathStr = headers.path.iterator.map(_.value).mkString("/")
+          RawRest.successfulAsync(RestResponse(404, HttpBody.plain(s"path $pathStr not found")))
+
+        case multiple =>
+          val pathStr = headers.path.iterator.map(_.value).mkString("/")
+          val callsRepr = multiple.iterator.map(p => s"  ${p.rpcChainRepr}").mkString("\n", "\n", "")
+          throw new RestException(s"path $pathStr is ambiguous, it could map to following calls:$callsRepr")
```

Comment:
I agree that it makes sense to refactor `resolvePath` to return `Opt`/`Option`

I don't understand why you want validation to be done lazily. I explicitly wanted it to be done eagerly so that it fails fast and can be put into a unit test. Ideally I'd like this validation to be done in compile time but that's impossible in this case.